### PR TITLE
[DEV 4198] Added logic for hopping and tumbling discrete windows in specs

### PR DESCRIPTION
**Pull request by nonibansal**

### CODE REVIEW

File: fennel/testing/execute_aggregation.py [hunk 1]
```diff
@@ -326,47 +329,240 @@ def get_val(self) -> List:
...
 
-        del_df = df[df[FENNEL_DELETE_TIMESTAMP].notna()]
-        del_df[FENNEL_ROW_TYPE] = -1
-        del_df[ts_field] = del_df[FENNEL_DELETE_TIMESTAMP]
-        df = pd.concat([df, del_df], ignore_index=True)
-        df = df.sort_values(
-            [ts_field, FENNEL_ROW_TYPE], ascending=[True, False]
-        )
+
+def postprocess_df(
```
Comment (aditya-nambiar) [inline fennel/testing/execute_aggregation.py:538]: where is this function being used ?

File: fennel/testing/execute_aggregation.py [hunk 2]
```diff
@@ -326,47 +329,240 @@ def get_val(self) -> List:
...
+
+
+def preprocess_df(
     input_df: pd.DataFrame,
     aggregate: AggregateType,
-    ts_field: str,
     key_fields: List[str],
-    output_dtype: Type,
+    ts_field: str,
 ) -> pd.DataFrame:
```
Comment (aditya-nambiar) [inline fennel/testing/execute_aggregation.py:392]: This function impl is so long. Can we break it up ?

File: fennel/testing/execute_aggregation.py [hunk 3]
```diff
@@ -326,47 +329,240 @@ def get_val(self) -> List:
         return self.vals[min(len(self.vals) - 1, max(index - 1, 0))]
 
 
-def get_aggregated_df(
+def get_timestamps_for_hopping_window(
```
Comment (aditya-nambiar) [inline fennel/testing/execute_aggregation.py:332]: Can we add tests for these functions ( get_timestamps_* ) too ?

File: fennel/testing/execute_aggregation.py [hunk 4]
```diff
@@ -326,47 +329,240 @@ def get_val(self) -> List:
...
+        if FENNEL_DELETE_TIMESTAMP in input_df.columns:
             del_df[ts_field] = del_df[[ts_field, FENNEL_DELETE_TIMESTAMP]].min(
                 axis=1
             )
-        df = pd.concat([df, del_df], ignore_index=True)
-        df = df.sort_values(
+        input_df = pd.concat([input_df, del_df], ignore_index=True)
+        input_df = input_df.sort_values(
             [ts_field, FENNEL_ROW_TYPE], ascending=[True, False]
         )
```
Comment (aditya-nambiar) [inline fennel/testing/execute_aggregation.py:534]: Lets move the common part of sorting and combining deletes out

File: fennel/testing/execute_aggregation.py [hunk 5]
```diff
@@ -326,47 +329,240 @@ def get_val(self) -> List:
...
-        # as a delete row
-        del_df = df.copy()
-        del_df[ts_field] = del_df[ts_field] + expire_secs
+
+        # Marking the delete timestamp as end timestamp of next session window.
+        fennel_internal_field = "__fennel_internal_field__"
+        internal_df = input_df.copy()[key_fields + [ts_field]].drop_duplicates()
+        internal_df[fennel_internal_field] = internal_df.groupby(key_fields)[
+            ts_field
+        ].shift(-1)
```
Comment (aditya-nambiar) [inline fennel/testing/execute_aggregation.py:516]: what does this do ? / add a comment

File: fennel/testing/execute_aggregation.py (hunk 1 above)
Comment (nonibansal) [reply fennel/testing/execute_aggregation.py:538]: Removed this

File: fennel/testing/execute_aggregation.py (hunk 2 above)
Comment (nonibansal) [reply fennel/testing/execute_aggregation.py:392]: broke up

File: fennel/testing/execute_aggregation.py (hunk 4 above)
Comment (nonibansal) [reply fennel/testing/execute_aggregation.py:534]: moved

File: fennel/testing/execute_aggregation.py (hunk 5 above)
Comment (nonibansal) [reply fennel/testing/execute_aggregation.py:516]: added

File: fennel/testing/execute_aggregation.py (hunk 3 above)
Comment (nonibansal) [reply fennel/testing/execute_aggregation.py:332]: added

File: fennel/client_tests/test_discrete_window_aggregation.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+                now_1d,
+                now_2d,
+                now_3d,
+                now_3d,
+                now_3d,
+            ],
+        }
+    )
+    client.log("fennel_webhook", "Transactions", df)
+    client.sleep()
```
Comment (aditya-nambiar) [inline fennel/client_tests/test_discrete_window_aggregation.py:71]: lets add the sleep when we convert this into a integration test

File: fennel/testing/execute_aggregation.py [hunk 7]
```diff
@@ -326,47 +329,256 @@ def get_val(self) -> List:
...
+    # of the two timestamps
+    if FENNEL_DELETE_TIMESTAMP in input_df.columns:
+        del_df[ts_field] = del_df[[ts_field, FENNEL_DELETE_TIMESTAMP]].min(
+            axis=1
+        )
+    input_df = pd.concat([input_df, del_df], ignore_index=True)
+    return input_df
+
+
+def preprocess_df_non_session_window(
```
Comment (aditya-nambiar) [inline fennel/testing/execute_aggregation.py:485]: s/preprocess_df_non_session_window(/preprocess_df_discrete_window

**Review by aditya-nambiar — APPROVED** [review]

File: fennel/client_tests/test_discrete_window_aggregation.py (hunk 6 above)
Comment (nonibansal) [reply fennel/client_tests/test_discrete_window_aggregation.py:71]: the sleep in normal client is zero

File: fennel/testing/execute_aggregation.py (hunk 7 above)
Comment (nonibansal) [reply fennel/testing/execute_aggregation.py]: changed

**Pull request merged**